### PR TITLE
Use builtin pandas functions to determine subsample groups

### PR DESCRIPTION
**Pull request by victorlin**
Optimize `get_groups_for_subsampling` with builtin pandas functions instead of inefficiently iterating through each row.

- [x] add tests
- [x] update code to pass tests
- [x] change logic to raise `FilterException` when `--group-by year month` is used without a `date` column (#795)

### Related issue(s)  

- #789 
    Note: With groupby and probabilistic sampling, I don't think there's a way to avoid full processing of the metadata file. Hopefully code optimization will result in enough performance improvement.
- #795

### Testing

- Added tests for `get_groups_for_subsampling`

### CODE REVIEW

**Comment by j23414** [comment]
> Note: With groupby and probabilistic sampling, I don't think there's a way to avoid full processing of the metadata file. 

I'm not sure if there are any [Reservoir sampling algorithms](https://link.springer.com/referenceworkentry/10.1007%2F978-0-387-39940-9_372) that would help. Might not be worth it depending on timing.

**Comment by victorlin** [comment]
Testing with 160k sequence dataset:

```sh
# 33.75s user 0.40s system 91% cpu 37.422 total
time augur filter \
  --metadata data/160k/metadata.tsv \
  --group-by country year month \
  --subsample-max-sequences 10 \
  --output-strains filter_strains.txt
```

If you want to look at profile results for yourself:

1. Download [160k-sequences-pr794.prof](https://gist.github.com/victorlin/e27255f0b4f833e04918fd0530bdc369/raw/7a358c4b5d4cce2bdfefe880fe739480483adf4f/160k-sequences-pr794.prof)
2. Go to https://nejc.saje.info/pstats-viewer.html. (SnakeViz web server hosted by a nice person on the internet)
3. Select Python 3.
4. Open downloaded file.

These changes bring the cumulative run time of `get_groups_for_subsampling` from [55s](https://github.com/nextstrain/augur/issues/789#issuecomment-984964323) down to 5s (row 3 column 2, need to hover on SnakeViz to see details).

snakeviz 160k-sequences-pr794.prof

**Comment by victorlin** [comment]
Note: there is a function [`pd.DataFrame.groupby`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.groupby.html) which I was expecting to use here, but ended up [dropping it](https://github.com/nextstrain/augur/commit/71291d92efc173e6a3a2c87184688904be97758a#diff-a5e96ae3f818746c03f1c454a53a66583c60fc867c90adda25a1c74403a77470L928) while fixing up the error handling. I suspect it will be useful for performance improvement in other parts of the code.

File: augur/filter.py [hunk 1]
```diff
@@ -884,91 +884,72 @@ def get_groups_for_subsampling(strains, metadata, group_by=None):
...
+            print(f"Filtering by group may behave differently than expected!", file=sys.stderr)
+            df_dates = pd.DataFrame({'year': 'unknown', 'month': 'unknown'}, index=metadata.index)
+            metadata = pd.concat([metadata, df_dates], axis=1)
+        else:
+            # replace date with year/month/day as nullable ints
+            date_cols = ['year', 'month', 'day']
+            df_dates = (metadata['date'].str.split('-', n=2, expand=True)
+                                            .set_axis(date_cols, axis=1))
+            for col in date_cols:
+                df_dates[col] = pd.to_numeric(df_dates[col], errors='coerce').astype(pd.Int64Dtype())
```
Comment (victorlin) [inline augur/filter.py:919]: Hmm, this is broken with incomplete dates (e.g. just `2020`). I'll write some tests and fix.

Comment (victorlin) [reply augur/filter.py:922]: False alarm – this is fine and there was another problem. Nevertheless, added a test here: https://github.com/nextstrain/augur/pull/794/commits/5596dd4bbcaab721b33b27206f6c1ef7e75a2bce

**Review by huddlej — APPROVED** [review]
Thank you for the thorough tests and elegant reimplementation of the grouping logic with pandas, @victorlin. I had a minor comment about unit test naming, but otherwise the code looks good. Also, I learned several pandas methods that I had previously missed, so thank you for that!

We haven’t had a chance to discuss team Git norms much, but we try to rebase branches on the `main` or `master` branch prior to merging, to avoid merge commits in the history. Rebasing is a little extra work, but it keeps the Git history linear and easier to follow. We also try to create highly descriptive commits that group logically-consistent changes (e.g., [the commits in this PR by @tsibley](https://github.com/nextstrain/cli/pull/115/commits)). I know GitHub provides a couple of options to automate this some process (either “Squash and merge” or “Rebase and merge”), but you may want to manually rebase, so you can squash specific groups of commits and edit their descriptions. For this PR, I’d consider squashing this to a single commit or one for tests and another for the code to make them pass.

I tested the runtimes of the following filter commands with and without the changes in this PR.

```
# Get 10 random sequences with "dummy" group.
# Ran in 2 min (same as pre-PR augur).
augur filter \
  --metadata metadata.tsv.gz \
  --query "division == 'Nebraska'" \
  --subsample-max-sequences 10 \
  --output-metadata nebraska_random_metadata.tsv \
  --output-strains nebraska_random_strains.txt

# Get 10 random sequences by month.
# Ran in 4 min (same as pre-PR augur)
augur filter \
  --metadata metadata.tsv.gz \
  --query "division == 'Nebraska'" \
  --subsample-max-sequences 10 \
  --group-by month \
  --output-metadata nebraska_random_metadata.tsv \
  --output-strains nebraska_random_strains.txt
```

These results seems consistent with [what you noticed in your own test of Trevor’s example command](https://github.com/nextstrain/augur/issues/789#issuecomment-985847112). Even if this PR doesn’t address the entire slow filter experience, it’s a step in the right direction for both code speed, readability, and maintainability. After you rebase/squash as desired, I’d plan to merge this.

Next, [we might want to cache the group results](https://github.com/nextstrain/augur/issues/789#issuecomment-985891027) and avoid the second calculation of these groups.

File: tests/test_filter_groupby.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+        group_by_strain, skipped_strains = get_groups_for_subsampling(strains, metadata, group_by=groups)
+        assert group_by_strain == {
+            'SEQ_1': ('A', 2020, (2020, 1)),
+            'SEQ_3': ('B', 2020, (2020, 3)),
+            'SEQ_4': ('B', 2020, (2020, 4)),
+            'SEQ_5': ('B', 2020, (2020, 5))
+        }
+        assert skipped_strains == [{'strain': 'SEQ_2', 'filter': 'skip_group_by_with_ambiguous_month', 'kwargs': ''}]
+
+    def test_filter_groupby_skip_month_2(self, valid_metadata: pd.DataFrame):
```
[the remark this answers is not in the document]
Comment (victorlin) [reply tests/test_filter_groupby.py]: Done: https://github.com/nextstrain/augur/pull/794/commits/2dcc8be7c23a193d647bdcbf3bd46584ee3ec796

This also made me think of another test, that an empty date would result in `skip_group_by_with_ambiguous_year`. Added here: https://github.com/nextstrain/augur/pull/794/commits/22230932277d3e709c09c3c98b437f28728b4b1c

**Comment by victorlin** [comment]
@huddlej re: Git norms, I started [a thread on Slack](https://bedfordlab.slack.com/archives/C01LCTT7JNN/p1636611620026200) and was able to get an answer from @ivan-aksamentov. I see that detailed individual commit messages are the norm on this team followed by `--no-ff` merge on PRs.

My workflow is iterative and I don't like getting bogged down in rebases/amending commits and rewriting working branch history. If it's OK, I'll use GitHub's **Squash and merge** feature to provide a detailed commit for `master`. This + modularized PRs should provide the same context that you all are used to seeing on commits to changes in the default branch.

**Comment by huddlej** [comment]
That solution sounds great for this kind of case where the PR is tightly scoped. For iterative work where PRs build on each other, we don't have as clearly defined standards. For example, during [the long-running augur filter refactor](https://github.com/nextstrain/augur/issues/699#issuecomment-875135042), @tsibley and I ended up preferring to close an earlier PR in favor of a later PR that built on the previous work.
